pathlib.Path support

d-chambers opened the pull request:
### What does this PR do?

This PR adds support for `pathlib.Path` for `read_events`, `read_inventory`, and `read` functions. 

It is implemented by simply converting `pathlib.Path` objects to str instances in `obspy.core.util.base._generic_reader`. 

### Why was it initiated?  Any relevant Issues?
It is onerous to have to convert `Path` objects to strings before reading inputs. This PR will just make it a bit more ergonomic. 

Support for `pathlib.Path` objects is discussed in #1794. However, that issue also talks about support for additional input types such as `ParseResult`, which I have not addressed here.

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] If the PR is making changes to documentation, docs pages can be built automatically.
      Just remove the space in the following string after the + sign: "+ DOCS"
- [ ] If any network modules should be tested for the PR, add them as a comma separated list
      (e.g. `clients.fdsn,clients.arclink`) after the colon in the following magic string: "+TESTS:"
      (you can also add "ALL" to just simply run all tests across all modules)
- [x] All tests still pass.
- [x] Any new features or fixed regressions are be covered via new tests.
- [x] Any new or changed features have are fully documented.
- [x] Significant changes have been added to `CHANGELOG.txt` .
- [x] First time contributors have added your name to `CONTRIBUTORS.txt` .

ThomasLecocq commented:
🎉 excellent! let me know if you need help !

d-chambers commented:
@ThomasLecocq,

Sure, feel free to review. It was surprisingly simple so maybe I am missing something but it seems to work fine.

ThomasLecocq commented:
> Sure, feel free to review. It was surprisingly simple so maybe I am missing something but it seems to work fine.

Well it depends if you want to use pathlib everywhere a path is used in the code :) there are quite a few, certainly in all the tests etc...

d-chambers commented:
> Well it depends if you want to use pathlib everywhere a path is used in the code :) there are quite a few, certainly in all the tests etc...

Although this would clean up the code quite a lot, its probably better left to be included in a larger refactoring effort. My main goal is to just ensure ObsPy's public functions which accept strings representing paths can also accept `pathlib.Path` objects. I only covered/tested `read`, `read_events`, and `read_inventory` so far but I probably need to track down others.

megies commented on obspy/core/tests/test_stream.py:
```diff
@@ -1948,6 +1948,16 @@ def test_read_url_via_network(self):
+        data_path = base_path / 'IU_ULN_00_LH1_2015-07-18T02.mseed'
+        assert data_path.exists()
+        st = read(data_path)
+        assert isinstance(st, Stream)
```
historically we used to go with unittest's `self.assertTrue()` but personally I don't mind the `assert`s, they're shorter and less convoluted and I don't think they should have any significant downsides as opposed to the unittest style

d-chambers replied:
@megies, your right. I have just become so accustomed to pytest I forgot to do that here, I will fix it.

megies commented on obspy/core/util/base.py:
```diff
@@ -646,6 +647,9 @@ def download_to_file(url, filename_or_buffer, chunk_size=1024):
                     **kwargs):
+    # convert pathlib.Path objects to str for compatibility.
+    if isinstance(pathname_or_url, Path):
+        pathname_or_url = str(pathname_or_url)
```
Do people actually use `PurePath`? I don't really know why they have that parent class to `Path` but since you are converting to string and then file opening is done outside of pathlib, we might just make this check go for `PurePath`, I guess?

d-chambers replied:
You are right, `PurePath` should be used here. If someone wanted to simulate path handling not native to the machine they are working on they have to reach for something besides `Path` (eg `WindowsPath` can be used on linux). However, in order for that to work we would need to wrap any instance of `PurePath` in a normal `Path` to make sure the string conversion will be in the right form.

megies replied:
My understanding from reading the docs is, `PurePath` is parent to `Path` implementing everything besides actual file opening. It's parent to all other Path flavors, so I would expect this is exactly what it's there for.

megies reviewed (approved):
One comment on the option to maybe use `PurePath`, otherwise looks good to go.

I think it would be a good idea to spell out the three user-faced functions that are getting pathlib-enabled by this.

Also, I think we want to mention this in the docs for these three functions, e.g. https://docs.obspy.org/master/packages/autogen/obspy.core.stream.read.html?highlight=pathname_or_url

megies commented:
such a small change only needed for this thanks to previous work by @ThomasLecocq in #2326 :clap:

d-chambers commented:
Alright @megies, I addressed your comments and performed a few small cleanups which were preventing the `pre-commit run -all` command from passing. Should be good to merge depending on CI.

megies commented on obspy/core/util/base.py:
```diff
@@ -646,6 +647,9 @@ def download_to_file(url, filename_or_buffer, chunk_size=1024):
                     **kwargs):
+    # convert pathlib.Path objects to str for compatibility.
+    if isinstance(pathname_or_url, PurePath):
+        pathname_or_url = str(Path(pathname_or_url))
```
No need to cast to `Path`, `__str__` is implemented by `PurePath` as parent anyway

d-chambers replied:
The reason I included this is a bit esoteric, but here is the weird edge case I had in mind:
```python
import pathlib
# assume we are using linux
windows_path = pathlib.PureWindowsPath('base') / 'filename.mseed'
print(str(windows_path))  # prints 'path\\filename.mseed' which is not a valid path for linux, hence we need the extra Path call
```
I don't think this will really happen much, if at all, but the code as it is written will handle it.

megies replied:
I still don't get it.. for me using..

```
print(str(pathlib.Path(windows_path)))
```

..in there doesn't change anything on the output

d-chambers replied:
Ok this is probably too far of an edge case anyway (why would someone use a PureWindowsPath on linux with ObSpy to read a file?) so I am going to just remove the extra `Path` constructor.

d-chambers commented:
rebased and force pushed

megies commented:
Thanks @d-chambers :rocket:

megies commented:
I just saw this @d-chambers: https://docs.python.org/3/library/os.html#os.PathLike
Not sure if that's probably even better, but I guess it might not make too much of a difference.

d-chambers commented:
Wow, I didn't know `PathLike` existed, I guess it pays to read the docs :stuck_out_tongue_winking_eye:. I guess `PathLike` would be better than `PurePath` but I think either one should work fine for how we are using them in this PR.
